File: python/817.linked-list-components.py
# Definition for singly-linked list.
from collections import defaultdict, deque
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def numComponents(self, head: Optional[ListNode], nums: List[int]) -> int:
        if not head:
            return 0

        prev, head = head, head.next
        to_visit = set(nums)
        res = len(nums)

        while head:
            if prev.val in to_visit and head.val in to_visit:
                res -= 1

            prev, head = head, head.next

        return res

File: golang/817.linked-list-components.go
/**
 * Definition for singly-linked list.
 */
package main

// type ListNode struct {
// 	Val  int
// 	Next *ListNode
// }

func numComponents(head *ListNode, nums []int) int {
	if head == nil || len(nums) == 0 {
		return 0
	}

	res := len(nums)

	toVisit := make(map[int]bool, res)
	for _, num := range nums {
		toVisit[num] = true
	}

	prev, cur := head, head.Next
	for cur != nil {
		if toVisit[prev.Val] && toVisit[cur.Val] {
			res--
		}
		prev, cur = cur, cur.Next
	}

	return res
}
